### ISSUE TITLE
Block registration from particular countries

### ISSUE THREAD

**Issue by ayub02** [openedx/platform-roadmap#390]
### Abstract

 Introduce a feature that allows platform administrators to restrict users from specific countries during sign-up and on account settings, addressing the need to control access based on legal, regulatory, or strategic requirements. Currently, there’s no mechanism for admins to restrict country access, requiring manual workarounds.

### Detailed Product Proposal

https://openedx.atlassian.net/wiki/spaces/OEPM/pages/4569825335/Proposal+Block+registration+from+particular+countries

### Context & Background (in brief, if a Product Proposal is linked above)

The Open edX platform doesn’t currently provide a way for administrators to restrict users from certain countries. For organizations facing legal restrictions, regulatory compliance, or strategic considerations, this feature is essential to manage user access seamlessly. Currently, the country list is retrieved separately across different micro-frontends (MFEs) on the web and via a backend API for mobile apps, which leads to inconsistencies in experience across platforms.

### Scope & Approach (in brief, if a Product Proposal is linked above)

The initiative focuses on modifying the country list based on a DISABLED_COUNTRIES setting, ensuring restricted countries are removed across all interfaces.

We have two implementation options:

Option 1: Maintain current structures by implementing backend filtering for both MFEs (via a new API) and mobile apps (via the registration API).
Option 2: Unify web and mobile by using a single registration API for country retrieval, which requires more extensive changes to MFEs but simplifies long-term maintenance.

### Value & Impact (in brief, if a Product Proposal is linked above)

Successful deployment of the DISABLED_COUNTRIES list without any technical errors across web and mobile platforms.

### Milestones and/or Epics

Option 1 is too small however option 2, if selected after review, can have milestones that are yet to be defined. 

### Named Release

Teak

### Timeline (in brief, if a Product Proposal is linked above)

Tentative plan is to implement as per available bandwidth. 

### Proposed By

2U

**Comment by sarina** [comment]
@ayub02 I want to touch base on this ticket, there is some conversation in the proposal about implementation. I think the product direction makes sense, but we need to iron out the implementation before moving forward with approval.

**Comment by ayub02** [comment]
@sarina sorry for late response. I want to set up a meet to discuss this with you and Asad. Can you please share your calendar where i can book some time for us?

**Comment by sarina** [comment]
@ayub02 - sure, you can make time here: https://calendly.com/sarina-tcril/30min

**Comment by marcotuts** [comment]
We had a version of this in use previously on the platform, and this is a rebuild or improvement? Previously (at least at edX) the country field was a required field in order to support blocking for certain courses etc. 

Just making sure others know that exists, but this seems like a more general / high level version of this tool.

**Comment by sarina** [comment]
@marcotuts yes we've discussed this on the wiki page.

@ayub02 any update on this?

**Comment by sarina** [comment]
@ayub02 no worries, I'm getting back from leave myself.

From a _product_ perspective, it looks like Marco took a look and I did as well, I think we can say this passes product review, provided that the old implementation _will_ be deprecated as indicated.

However you should be sure to run your proposed implementation by the relevant technical teams (frontend and mobile) before implementation - I am not a frontend or mobile developer, so I'm not qualified to comment on the substance of your proposed implementation.

**Comment by awais-ansari** [comment]
Hello @sarina, Could you please direct us to the appropriate frontend and mobile teams or channels to discuss this proposal? Thanks!

**Comment by sarina** [comment]
On the Open edX side the rooms are called wg-frontend and mobile-apps. I do not know what they are called in your private 2U slack.

**Comment by regisb** [comment]
Is the goal to implement this feature using the Open edX events and hooks framework? It seems like our extensibility framework was created precisely to address this kind of use cases.

**Comment by awais-ansari** [comment]
> Is the goal to implement this feature using the Open edX events and hooks framework? It seems like our extensibility framework was created precisely to address this kind of use cases.

Technically yes we can add new hooks for this but This feature is something that is a common use case that the wider community will benefit from, So building it like a platform feature makes more sense.

**Comment by AhtishamShahid** [comment]
> > This feature is something that is a common use case that the wider community will benefit from, So building it like a platform feature makes more sense.
> 
> I find this surprising. It's the first time in my career that I hear about such a feature in a web app. Could we have an opinion from product here? To clarify: the question is whether this feature should be implemented in edx-platform (Open edX "core") or as a pluggable feature, for instance using the hooks extensibility framework.

The embargo feature already exists in edx-platfrom for a similar use case (the Embargo app currently supports blocking country access to a single course ). This proposal will extend the existing implementation of the embargo so staff members can block countries globally .
Now the question is, should we use the existing embargo app to implement this feature or use Hooks for this? In my opinion, extending the existing emargo app will make more sense because it aligns with the current design and avoids duplicating functionality. The embargo app is already responsible for access restrictions, so extending it to support global blocking would ensure consistency.

**Comment by sarina** [comment]
@regisb Embargo is a need for many US-based as well as governmental organizations complying with embargo/access restriction rules enforced by their countries. I know for certain both Axim and 2U are bound by such embargo laws.

**Comment by regisb** [comment]
Ha OK, I wasn't aware of that, my bad. And given the current political context, we should probably expect more and more embargo embargo laws in the near term.

Still, I'd like to raise the question: what qualifies a feature to be built into edx-platform core, as opposed to building it as a pluggable extension? (using for instance the hooks framework)

**Comment by sarina** [comment]
It's a good question. My understanding from the document is that the existing functionality will be deprecated in favor of this implementation, rather than augmenting the existing functionality. In that case, if this _can_ be implemented by hooks, I think it ought to be. @ormsbee do you have any strong opinions?

**Comment by ayub02** [comment]
@sarina @ormsbee PRs for this have been ready to merge for several weeks now. If there are no immediate concerns, can we merge these changes?

**Comment by ormsbee** [comment]
Yes, it's fine to merge as part of edx-platform. Maybe it makes sense to move into a more general access control related set of hooks later, but I don't think it should block this now. My apologies for the delayed response.

**Comment by ayub02** [comment]
Thank you @ormsbee @sarina and @regisb and @marcotuts for spending time reviewing this.

@AhtishamShahid i believe we are good to proceed with merging PRs and testing.

**Comment by AhtishamShahid** [comment]
Thank you everyone for your insights. We will merge the changes on Monday.

**Comment by sarina** [comment]
@AhtishamShahid can this be closed?

**Comment by AhtishamShahid** [comment]
Hi, PRs were merged and tested successfully, Closing this now.
